File: javascript/packages/linter/src/rules/actionview-no-render-option-shadowing.ts
import { BaseRuleVisitor } from "./rule-utils.js"
import { ParserRule } from "../types.js"

import { isPrismNodeType, isRubyRenderLocalNode } from "@herb-tools/core"

import type { ERBRenderNode, ParseResult, ParserOptions, RubyRenderLocalNode } from "@herb-tools/core"
import type { FullRuleConfig, LintContext, UnboundLintOffense } from "../types.js"

const RENDER_OPTIONS = new Set([
  "as",
  "cached",
  "collection",
  "content_type",
  "formats",
  "handlers",
  "layout",
  "object",
  "spacer_template",
  "template",
  "variants",
])

class ActionViewNoRenderOptionShadowingVisitor extends BaseRuleVisitor {
  visitERBRenderNode(node: ERBRenderNode): void {
    this.checkRender(node)

    this.visitChildNodes(node)
  }

  private checkRender(node: ERBRenderNode): void {
    if (!this.isShorthandForm(node)) return

    for (const local of node.keywords?.locals ?? []) {
      if (!isRubyRenderLocalNode(local)) continue

      this.checkLocal(local)
    }
  }

  private isShorthandForm(node: ERBRenderNode): boolean {
    const call = node.prismNode

    if (!isPrismNodeType(call, "CallNode")) return false

    const [first] = call.arguments_?.arguments_ ?? []

    if (!first) return false

    return !isPrismNodeType(first, "KeywordHashNode")
  }

  private checkLocal(local: RubyRenderLocalNode): void {
    const name = local.name?.value

    if (!name) return
    if (!RENDER_OPTIONS.has(name)) return

    this.addOffense(
      `The local \`${name}\` shadows the \`render\` option of the same name. Every keyword in the shorthand \`render "partial", ...\` form becomes a local, but the same keyword after \`render partial: "..."\` is a render option instead. Move it into an explicit \`locals: { ${name}: ... }\` hash so there is only one way to read it.`,
      local.name!.location,
    )
  }
}

export class ActionViewNoRenderOptionShadowingRule extends ParserRule {
  static ruleName = "actionview-no-render-option-shadowing"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "info",
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      render_nodes: true,
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ActionViewNoRenderOptionShadowingVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/actionview_no_render_option_shadowing.rs
use crate::utils::prism_utils::call_arguments;

use herb::nodes::{AnyNode, ERBRenderNode, RubyRenderLocalNode};
use herb::Visitor;

rule_visitor!(NoRenderOptionShadowingVisitor);
define_parser_rule!(
  ActionViewNoRenderOptionShadowingRule,
  "actionview-no-render-option-shadowing",
  Info,
  NoRenderOptionShadowingVisitor,
  parser_options: { render_nodes: true, prism_nodes: true },
  introduced_in: "unreleased"
);

const RENDER_OPTIONS: &[&str] = &[
  "as",
  "cached",
  "collection",
  "content_type",
  "formats",
  "handlers",
  "layout",
  "object",
  "spacer_template",
  "template",
  "variants",
];

fn is_shorthand_form(node: &ERBRenderNode) -> bool {
  let call = match node.prism() {
    Some(call) if call.is("CallNode") => call,
    _ => return false,
  };

  match call_arguments(call).first() {
    Some(first) => !first.is("KeywordHashNode"),
    None => false,
  }
}

impl NoRenderOptionShadowingVisitor {
  fn check_local(&mut self, local: &RubyRenderLocalNode) {
    let name = match local.name.as_ref() {
      Some(name) => name,
      None => return,
    };

    if !RENDER_OPTIONS.contains(&name.value.as_str()) {
      return;
    }

    self.add_offense(
      format!(
        "The local `{}` shadows the `render` option of the same name. Every keyword in the shorthand `render \"partial\", ...` form becomes a local, but the same keyword after `render partial: \"...\"` is a render option instead. Move it into an explicit `locals: {{ {}: ... }}` hash so there is only one way to read it.",
        name.value, name.value
      ),
      name.location.clone(),
    );
  }
}

impl Visitor for NoRenderOptionShadowingVisitor {
  fn visit_erb_render_node(&mut self, node: &ERBRenderNode) {
    if is_shorthand_form(node) {
      let locals: Vec<&RubyRenderLocalNode> = node
        .keywords
        .as_ref()
        .map(|keywords| {
          keywords
            .locals
            .iter()
            .filter_map(|local| match local {
              AnyNode::RubyRenderLocalNode(local) => Some(local.as_ref()),
              _ => None,
            })
            .collect()
        })
        .unwrap_or_default();

      for local in locals {
        self.check_local(local);
      }
    }

    self.walk_erb_render_node(node);
  }
}
